runtime/parachains/src/disputes/slashing.rs

```diff
@@ -0,0 +1,905 @@
+		key_owner_proof: T::KeyOwnerProof,
+	) -> DispatchResult;
+}
+
+impl<T: Config> HandleReports<T> for () {
+	type ReportLongevity = ();
+
+	fn report_for_invalid_offence(
+		_reporters: Vec<AccountId<T>>,
+		_offence: ForInvalidOffence<T::KeyOwnerIdentification>,
+	) -> Result<(), OffenceError> {
+		Ok(())
+	}
+
+	fn report_against_valid_offence(
+		_reporters: Vec<AccountId<T>>,
+		_offence: AgainstValidOffence<T::KeyOwnerIdentification>,
+	) -> Result<(), OffenceError> {
+		Ok(())
+	}
+
+	fn is_known_for_invalid_offence(
+		_offenders: &[T::KeyOwnerIdentification],
+		_time_slot: &DisputesTimeSlot,
+	) -> bool {
+		true
+	}
+
+	fn is_known_against_valid_offence(
+		_offenders: &[T::KeyOwnerIdentification],
+		_time_slot: &DisputesTimeSlot,
+	) -> bool {
+		true
+	}
+
+	fn submit_unsigned_slashing_report(
+		_dispute_proof: DisputeProof,
+		_key_owner_proof: T::KeyOwnerProof,
+	) -> DispatchResult {
+		Ok(())
+	}
+}
+
+pub trait WeightInfo {
+	fn report_dispute_lost(validator_count: ValidatorSetCount) -> Weight;
+}
+
+pub struct TestWeightInfo;
+impl WeightInfo for TestWeightInfo {
+	fn report_dispute_lost(_validator_count: ValidatorSetCount) -> Weight {
+		0
+	}
+}
+
+pub use pallet::*;
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + crate::disputes::Config {
+		/// The proof of key ownership, used for validating slashing reports.
+		/// The proof must include the session index and validator count of the
+		/// session at which the offence occurred.
+		type KeyOwnerProof: Parameter + GetSessionNumber + GetValidatorCount;
+
+		/// The identification of a key owner, used when reporting slashes.
+		type KeyOwnerIdentification: Parameter;
+
+		/// A system for proving ownership of keys, i.e. that a given key was part
+		/// of a validator set, needed for validating slashing reports.
+		type KeyOwnerProofSystem: KeyOwnerProofSystem<
+			(KeyTypeId, ValidatorId),
+			Proof = Self::KeyOwnerProof,
+			IdentificationTuple = Self::KeyOwnerIdentification,
+		>;
+
+		/// The slashing report handling subsystem, defines methods to report an
+		/// offence (after the slashing report has been validated) and for submitting a
+		/// transaction to report a slash (from an offchain context).
+		/// NOTE: when enabling slashing report handling (i.e. this type isn't set to
+		/// `()`) you must use this pallet's `ValidateUnsigned` in the runtime
+		/// definition.
+		type HandleReports: HandleReports<Self>;
+
+		/// Weight information for extrinsics in this pallet.
+		type WeightInfo: WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	/// Indices of the validators pending "for invalid" dispute slashes.
+	#[pallet::storage]
+	pub(super) type PendingForInvalidLosers<T> =
+		StorageDoubleMap<_, Twox64Concat, SessionIndex, Blake2_128Concat, CandidateHash, Losers>;
```

Comment:
why re-hash `CandidateHash`?

Reply:
(questioning your choice of `Blake2_128Concat` here)

Reply:
Using the same hash as in https://github.com/paritytech/polkadot/blob/e7461a6bd020250fed010ba56b78b854ad463064/runtime/parachains/src/disputes.rs#L430-L439

My understanding is `CandidateHash` can be potentially manipulated by malicious collators, so we can rely on the uniformity of the distribution.